pkg/virt-controller/services/template.go

```diff
@@ -144,6 +144,31 @@ func isSRIOVVmi(vmi *v1.VirtualMachineInstance) bool {
 	return false
 }
 
+func requestResource(resources *k8sv1.ResourceRequirements, resourceName string) {
+	name := k8sv1.ResourceName(resourceName)
+
+	// assume resources are countable, singular, and cannot be divided
+	unitQuantity := *resource.NewQuantity(1, resource.DecimalSI)
+
+	// Fill in limits
+	val, ok := resources.Limits[name]
+	if ok {
+		val.Add(unitQuantity)
+		resources.Limits[name] = val
+	} else {
+		resources.Limits[name] = unitQuantity
+	}
+
+	// Fill in requests
+	val, ok = resources.Requests[name]
+	if ok {
+		val.Add(unitQuantity)
+		resources.Requests[name] = val
+	} else {
+		resources.Requests[name] = unitQuantity
```

Comment:
Usually, it should be fine to avoid setting Requests for a certain resource if a corresponding Limits is set for it.
When only Limits is set for a certain resource k8s will make it be equal to resource's Requests.
If all Limits are equal to Requests (or Requests not set at all) Pod's QOS will be set to `Guaranteed`.
As it is now if the user has set only Limits, we will set Requests to `unitQuantity`, which will affect Pod's QOS... 
I think we only need to set Requests for a certain resource if it's missing Limits.

Reply:
@vladikr I will need to think about it, but this logic repeats what was proposed for Multus webhook. If there is an issue in this logic we should bring it up to Multus folks to address too. Let me think about it.

Reply:
@vladikr from device plugin design document: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md#end-user-story

```
Devices can be selected using the same process as for OIRs in the pod spec. Devices have no impact on QOS. However, for the alpha, we expect the request to have limits == requests.
```

This wording suggests that we should always set limits == requests for devices managed by a device plugin (for the alpha; the feature is in beta right now). I am not aware of any changes that happened since then in regards to device plugin API that would suggest a change for limits management.

Reply:
@booxter Yeah, I've expected to see limits == requests. Great that it doesn't affect QOS.

But right now, I see that the code above allows setting it differently.
If limits = 2 and no requests set  - requests will be set to 1, no?

Reply:
@vladikr yes but that would a problem of those who put incorrect limit that doesn't match requests in the first place.

Reply:
@booxter I'm not sure, I think normally in k8s if a user specifies only limits it implies that requests will match the limits.
This is how it works with cpu and mem, at least.[1]
`Note: If a Container specifies its own memory limit, but does not specify a memory request, Kubernetes automatically assigns a memory request that matches the limit. Similarly, if a Container specifies its own CPU limit, but does not specify a CPU request, Kubernetes automatically assigns a CPU request that matches the limit.`

I don't have a strong opinion about this. If you think that the current behavior will not surprise the users, I'm fine with it. 

[1] https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed

Reply:
@vladikr I don't think we should refer to QoS documentation for a feature that explicitly says in its design document that it doesn't affect QoS.

But let's run the idea. How would the implementation look like? I will try something in pseudocode, please tell me if this is what you would have here.

```
func requestResource(resources, resourceName) {
    if resourceName in resources.Limits and resourceName not in resources.Requests {
        resources.Requests[resourceName] = resources.Limits[resourceName]
    }
    resources.Requests[resourceName]++
    resources.Limits[resourceName]++
}
```

Reply:
:) something like that.

But, I think I would just change 
`} else {
    resources.Requests[name] = unitQuantity
`

to something like: 
```
} else {
    val, _ := resources.Limits[name]
    resources.Requests[name] = val
}
```
Since the code already deals with `Limits` we can just set the limits value for the missing `Requests`

Reply:
> I don't think we should refer to QoS documentation for a feature that explicitly says in its design document that it doesn't affect QoS.

@booxter You're right. I've just seen it in other places as well, like `LimitRange` - but it was also realtred to cpu/mem I think.